CSS Nesting: fix performance regression by storing selector in StyleRule

mdubet opened the pull request:
#### 29a55829552726ea8b7310c95c3ff4bbcf591583
<pre>
CSS Nesting: fix performance regression by storing selector in StyleRule
<a href="https://bugs.webkit.org/show_bug.cgi?id=252904">https://bugs.webkit.org/show_bug.cgi?id=252904</a>
rdar://105883716

Reviewed by Antti Koivisto.

CSS Nesting introduces a performance regression in selector
matching because of a branch to select the appropriate selector list
(user-written/nested one VS resolved/flatten one).

This patch stores the resolved selector list in the original StyleRule,
thus allowing to have only this object to perform selector matching.

The original user-written selector list is stored in the StyleRuleWithNesting
object which is not an issue because it&apos;s never in a hot code path.

* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::generateSelectorText const):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleWithNesting::StyleRuleWithNesting):
(WebCore::StyleRule::resolvedSelectorList const): Deleted.
(WebCore::StyleRuleWithNesting::setResolvedSelectorList const): Deleted.
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::buildArrayForMatchedRuleList):
* Source/WebCore/style/RuleData.h:
(WebCore::Style::RuleData::selector const):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::resolveSelectorListWithNesting):
(WebCore::Style::RuleSetBuilder::addStyleRule):
(WebCore::Style::RuleSetBuilder::populateStyleRuleResolvedSelectorList): Deleted.
* Source/WebCore/style/RuleSetBuilder.h:

Canonical link: <a href="https://commits.webkit.org/261809@main">https://commits.webkit.org/261809@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aef2cdd71e476471e03cb22dde104783623e97e9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112932 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/22090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1607 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4708 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/121428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/117036 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23437 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13251 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/5900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118719 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100665 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106021 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/1203 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46427 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14385 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1243 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15090 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20401 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53235 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8243 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16941 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/RuleSetBuilder.cpp:
```diff
@@ -215,13 +215,14 @@ void RuleSetBuilder::addRulesFromSheetContents(const StyleSheetContents& sheet)
 }
 
-void RuleSetBuilder::populateStyleRuleResolvedSelectorList(const StyleRuleWithNesting& rule)
+void RuleSetBuilder::computeUsableSelectorList(StyleRuleWithNesting& rule)
```
"Usable" seems like a new concept and it is not clear what it means (or is it a spec term?). How about something like `resolveSelectorListWithNesting`?

mdubet replied:
No it's a made up term by myself. I will change to use `resolveSelectorListWithNesting`.

anttijk commented on Source/WebCore/style/RuleSetBuilder.h:
```diff
@@ -39,7 +39,7 @@ class RuleSetBuilder {
     RuleSetBuilder(const MQ::MediaQueryEvaluator&);
 
-    void addStyleRule(const StyleRuleWithNesting&);
+    void addStyleRule(StyleRuleWithNesting&);
```
`addStyleRuleWIthNesting`?

mdubet replied:
In a precedent review, @darinadler commented that we should let the C++ compiler do the dispatching and keep the name clear of the type. I don't have a strong opinion on this (actually I started this patch with `addStyleRuleWithNesting` but then remembered this old comment 😃 )

darinadler replied:
If the longer name is the most logical, then I think it’s OK to use it. But if the shorter name makes sense, I think it is indeed better to not repeat the names of argument types in a language that allows us to overload.

mdubet committed:
CSS Nesting: fix performance regression by storing selector in StyleRule
https://bugs.webkit.org/show_bug.cgi?id=252904
rdar://105883716

Reviewed by Antti Koivisto.

CSS Nesting introduces a performance regression in selector
matching because of a branch to select the appropriate selector list
(user-written/nested one VS resolved/flatten one).

This patch stores the resolved selector list in the original StyleRule,
thus allowing to have only this object to perform selector matching.

The original user-written selector list is stored in the StyleRuleWithNesting
object which is not an issue because it's never in a hot code path.

* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::generateSelectorText const):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleWithNesting::StyleRuleWithNesting):
(WebCore::StyleRule::resolvedSelectorList const): Deleted.
(WebCore::StyleRuleWithNesting::setResolvedSelectorList const): Deleted.
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::buildArrayForMatchedRuleList):
* Source/WebCore/style/RuleData.h:
(WebCore::Style::RuleData::selector const):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::resolveSelectorListWithNesting):
(WebCore::Style::RuleSetBuilder::addStyleRule):
(WebCore::Style::RuleSetBuilder::populateStyleRuleResolvedSelectorList): Deleted.
* Source/WebCore/style/RuleSetBuilder.h:

Canonical link: https://commits.webkit.org/261809@main
